lib/ansible/modules/cloud/smartos/vmadm.py

```diff
@@ -0,0 +1,609 @@
+        supports_check_mode=True,
+        required_one_of=[['name', 'uuid']]
+    )
+
+    module.vmadm = module.get_bin_path('vmadm', required=True)
+
+    p = module.params
+    uuid = p['uuid']
+    state = p['state']
+
+    # Translate the state paramter into something we can use later on.
+    if state in ['present', 'running']:
+        vm_state = 'running'
+    elif state in ['stopped', 'created']:
+        vm_state = 'stopped'
+    elif state in ['absent', 'deleted']:
+        vm_state = 'deleted'
+    elif state in ['restarted', 'rebooted']:
+        vm_state = 'rebooted'
+
+    result = {'state': state}
+    changed = False
+
+    # While it's possible to refer to a given VM by it's `alias`, it's easier
+    # to operate on VMs by their UUID. So if we're not given a `uuid`, look
+    # it up.
+    if not uuid:
+        uuid = get_vm_uuid(module, p['name'])
+        # Bit of a chicken and egg problem here for VMs with state == deleted.
+        # If they're going to be removed in this play, we have to lookup the
+        # uuid. If they're already deleted there's nothing to looup.
+        # So if state == deleted and get_vm_uuid() returned '', the VM is already
+        # deleted and there's nothing else to do.
+        if uuid == '' and vm_state == 'deleted':
+            result['name'] = p['name']
+            module.exit_json(**result)
+
+    validate_uuids(module)
+
+    if p['name']:
+        result['name'] = p['name']
+    result['uuid'] = uuid
+
+    if uuid == '*':
+        result['changed'] = manage_all_vms(module, vm_state, changed)
+        module.exit_json(**result)
+
+    # The general flow is as follows:
+    # - first the current state of the VM is obtained by it's UUID.
+    # - If the state was not found and the desired state is 'deleted', return.
+    # - If the state was not found, it means the VM has to be created.
+    #   Subsequently the VM will be set to the desired state (i.e. stopped)
+    # - Otherwise, it means the VM exists already and we operate on it's
+    #   state (i.e. reboot it.)
+    #
+    # In the future it should be possible to query the VM for a particular
+    # property as a valid state (i.e. queried) so the result can be
+    # registered.
+    # Also, VMs should be able to get their properties updated.
+    # Managing VM snapshots should be part of a standalone module.
+
+    # First obtain the VM state to determine what needs to be done with it.
+    current_vm_state = get_vm_prop(module, uuid, 'state')
+
+    # First handle the case where the VM should be deleted and is not present.
+    if not current_vm_state and vm_state == 'deleted':
+        changed = False
+    elif module.check_mode:
+        # Shortcut for check mode, if there is no VM yet, it will need to be created.
+        # Or, if the VM is not in the desired state yet, it needs to transition.
+        if (not current_vm_state) or (get_vm_prop(module, uuid, 'state') != state):
+            result['changed'] = True
```

Comment:
Either use `result['changed'] = ` and remove the `changed =` OR use `changed =`  and remove the `result['changed']` but do not use both. That applies to all `main()` function.